### PR TITLE
AoC 2019 Day 10 "Asteroids, normalisation, trig"

### PR DESCRIPTION
**Pull request by P-on-D**
Not proud of the code for this but I am pleased I got a "mathematical" solution. Took a lot of fumbling about so I kept my working.

### CODE REVIEW

File: 2019/d10Asteroids.d [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+auto vapourisationOrder(Pt[] asteroids, Pt base) {
+  import std.math;
+
+  Pt[] remaining = asteroids.filter!(a => a != base).array.sort.array;
+  Pt[] candidates;
+
+  do {
+    auto visible = remaining.visibleFrom(base);
+
+    candidates ~= visible
```
Comment (P-on-D-Reviewer) [inline 2019/d10Asteroids.d]: If you write out in a comment what each step is doing here, you'll be able to do a much better job of implementing the algorithm. You haven't used `xdir` and `ydir` for instance.

File: 2019/d10Asteroids.d [hunk 2]
```diff
@@ -0,0 +1,291 @@
+import std.array, std.algorithm, std.typecons;
+
+struct Pt { int x, y; int opCmp(ref const Pt s) const { return x == s.x ? y - s.y : x - s.x; } }
+
+struct NMV {
```
Comment (P-on-D-Reviewer) [inline 2019/d10Asteroids.d:25]: Generally, it would be excellent practice to comment this code as it isn't obvious what's going on. Pretty please?

**Review by P-on-D-Reviewer — CHANGES_REQUESTED** [review]

File: 2019/d10Asteroids.d (hunk 1 above)
Comment (P-on-D) [reply 2019/d10Asteroids.d]: That worked well. I need to focus on how to work better with ranges and avoid the `.array` spam.

**Review by P-on-D-Reviewer — APPROVED** [review]
Better. Still some awkwardness to be worked on.

**Pull request merged**